core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,136 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.ValidationException;
+
+/**
+ * In an audit workflow, new data is written to an orphan snapshot that is not committed as the table's
+ * current state until it is audited. After auditing a change, it may need to be applied or cherry-picked
+ * on top of the latest snapshot instead of the one that was current when the audited changes were created.
+ *
+ * This class adds support for cherry-picking the changes from an orphan snapshot by applying them to
+ * the current snapshot. The output of the operation is a new snapshot with the changes from cherry-picked
+ * snapshot.
+ *
+ * Cherry-picking should apply the exact set of changes that were done in the original commit.
+ *  - All added files should be added to the new version.
+ *  - Todo: If files were deleted, then those files must still exist in the data set.
+ *  - Does not support Overwrite operations currently. Overwrites are considered as conflicts.
+ *
+ */
+class CherryPickFromSnapshot extends MergingSnapshotProducer<CherryPick> implements CherryPick {
+  private final TableOperations ops;
+  private TableMetadata base;
+  private Long cherryPickSnapshotId = null;
+
+  CherryPickFromSnapshot(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.refresh();
+  }
+
+  @Override
+  protected CherryPick self() {
+    return this;
+  }
+
+  /**
+   * We only cherry pick for appends right now
+   * @return
+   */
+  @Override
+  protected String operation() {
+    Snapshot cherryPickSnapshot = base.snapshot(cherryPickSnapshotId);
+    return cherryPickSnapshot.operation();
+  }
+
+  @Override
+  public CherryPickFromSnapshot cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    this.cherryPickSnapshotId = snapshotId;
+    return this;
+  }
+
+  /**
+   * Apply the pending changes and return the uncommitted changes for validation.
+   * <p>
+   * This does not result in a permanent update.
+   *
+   * @return the uncommitted changes that would be committed by calling {@link #commit()}
+   * @throws ValidationException      If the pending changes cannot be applied to the current metadata
+   * @throws IllegalArgumentException If the pending changes are conflicting or invalid
+   */
+  @Override
+  public Snapshot apply() {
```

Comment:
I think it would be easier overall if you didn't override apply. Then you also wouldn't need to override commit and you'd get retries for free.

Reply:
done